File: GFG/Hamiltonian Path/Hamiltonian Path.cpp
class Solution
{
    public:
    bool dfs(int n, vector<int> &visited, vector<int> adj[],int node){
        if(n == 0) return true;
        visited[node] = 1;
        for(auto nbr: adj[node]){
            if(visited[nbr] == 0){
             if(dfs(n-1, visited, adj, nbr)){
                return true;
              }
            }
        }
        visited[node] = 0;
        return false;
    }
    public:
    bool check(int N,int M,vector<vector<int>> Edges)
    {
       vector<int> adj[N];
       for(auto i: Edges){
           adj[ i[0] - 1].push_back(i[1] - 1);
           adj[ i[1] - 1].push_back(i[0] - 1);
       }
       vector<int> visited(N, 0);
       
       for(int i = 0;i<N;i++){
           if(dfs(N-1, visited, adj, i)){
               return true;
           }
       }
       return false;
    }
};

File: GFG/Hamiltonian Path/Hamiltonian Path.java
class Solution 
{ 
    int n;
    boolean check(int N, int M, ArrayList<ArrayList<Integer>> Edges) 
    { 
        // code here
        n=N;
        Map<Integer,ArrayList<Integer>> map=new HashMap<>();
        for(ArrayList<Integer> edge:Edges){
            int src=edge.get(0);
            int des=edge.get(1);
            map.putIfAbsent(src,new ArrayList<>());
            map.putIfAbsent(des,new ArrayList<>());
            map.get(src).add(des);
            map.get(des).add(src);
        }
        
        Set<Integer> set=new HashSet<>();
        for(int node=1;node<=N;node++){
            if(dfs(node,set,map)){
                return true;
            }
        }
        return false;
    }
    boolean dfs(int curr,Set<Integer> set,Map<Integer,ArrayList<Integer>> map){
        set.add(curr);
        if(set.size()==n){
            return true;
        }
        
        for(int neb:map.get(curr)){
            if(!set.contains(neb) && dfs(neb,set,map)){
                return true;
            }
        }
        
        set.remove(curr);//back..
        return false;
    }
} 
